server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java

```diff
@@ -44,6 +45,12 @@ private void expand() {
 
     public void remove() {
         path[--index] = null;
+
+        // Reset the StringBuilder if it includes a newly removed field
```

Comment:
I think we can avoid a full reset here. We can determine the length of the removed path element before nulling it out, then reset the string builder length accordingly. Some special casing for index == 0 and the delimiters might be needed.